[CORE-8690]: Update offline log viewer to be able to parse crash reports

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
crash_tracker: Create crash_report_generator

Created a simple CLI tool that can be used to generate crash reports on
demand.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
viewer: Added support for parsing crash reports

offline_log_viewer now supports decoding crash reports in two ways:

* Directly pointing the script at a crash report file
* Pointing the script at the Redpanda data directory and parsing every
  crash report in the crash_reports directory

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
crash_tracker: Change underlying type in reserved_string

Switch from `ss::sstring` to `std::array<char, Size>`.  This provides
stronger guarantees that the reserved string buffer does not change size
or attempts to perform operations on the heap (like free or new) while
operating in a signal handler.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt/clients: Updated offline_log_viewer

Updated to support reading crash_reports directory

Signed-off-by: Michael Boquard <michael@redpanda.com>

pgellert commented on src/v/crash_tracker/types.h:
```diff
@@ -30,38 +30,72 @@ enum class crash_type {
     // NOLINTNEXTLINE(hicpp-explicit-conversions)
-    reserved_string(ss::sstring other) noexcept
-      : _str(std::move(other)) {};
+    explicit reserved_string(ss::sstring other) noexcept {
```
I think that `NOLINTNEXTLINE` is no longer necessary

pgellert commented on src/v/crash_tracker/types.h:
```diff
@@ -72,23 +106,21 @@ struct crash_description
-      , crash_message(string_buffer_reserve)
-      , stacktrace(string_buffer_reserve)
-      , addition_info(string_buffer_reserve) {}
+      : crash_time{} {}
```
nit: I think this constructor could now be undeclared / `default`ed

pgellert commented on tests/rptest/clients/offline_log_viewer.py:
```diff
@@ -55,6 +55,9 @@ def read_controller_snapshot(self, node):
     def read_consumer_offsets(self, node):
         return self._json_cmd(node, "--type consumer_offsets")
 
+    def read_crash_reports(self, node):
+        return self._json_cmd(node, "--type crash_report")
```
Should we add a test in `crash_loop_checks_test.py` that uses this? Either by adding a check at the end of `test_crash_report_with_startup_exception` or in a new short test that just records a startup crash and reads it back.

pgellert commented on src/v/crash_tracker/report_generator.cc:
```diff
@@ -0,0 +1,92 @@
+#include <exception>
+
+namespace {}
+
```
nit: unused
```suggestion
```

pgellert commented on src/v/crash_tracker/report_generator.cc:
```diff
@@ -0,0 +1,92 @@
+
+namespace {}
+
+int main(int ac, char* av[]) {
```
Nice!

pgellert commented on src/v/crash_tracker/types.h:
```diff
@@ -30,38 +30,72 @@ enum class crash_type {
+  serde::tag_t<serde::write_tag>, iobuf& out, reserved_string<Size> t) {
+    static_assert(Size <= std::numeric_limits<serde::serde_size_t>::max());
+    serde::write(out, static_cast<serde::serde_size_t>(t.length()));
+    auto begin = t.begin();
+    for (size_t i = 0; i < t.length(); ++i) {
```
Is it worth adding a comment here to call out that the only difference between this and the serde of `array` is that we only serialize the first `length()` chars? (Afaict that's the only difference.)
```suggestion
    auto begin = t.begin();
    // Note: we only serialize the first `length()` characters of the string and not the full `capacity()`
    for (size_t i = 0; i < t.length(); ++i) {
```

michael-redpanda commented:
Force push [df59bcc](https://github.com/redpanda-data/redpanda/commit/df59bccf21932ab5597252d7660e459ddba9d7c3):

* Fixed bazel lint error

pgellert commented on src/v/crash_tracker/types.h:
```diff
@@ -30,38 +30,72 @@ enum class crash_type {
-  : serde::
-      envelope<reserved_string, serde::version<0>, serde::compat_version<0>> {
+template<size_t Size>
+struct reserved_string {
+    using underlying_t = std::array<char, Size>;
```
I think at the moment we could fill up the full `Size` characters with "content" and then we loose the nul-termination that's needed at least for `length()`.

Maybe the easiest option is for the array to have a length of `Size + 1` and `capacity()` to return `Size`?

pgellert reviewed (commented):
looks great

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e247df9d-c5ba-4b91-b731-c63483b4cc10-->
<!--TEST RESULTS-->
<!--b55e32c3-0caa-4a0a-a637-fbec8b5674f1-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61370">build#61370</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61370#0194b37a-6489-4b16-adc8-d9cd58b60ffa | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61370#0194b3bc-c936-42ec-bb28-59fedc8426af | FLAKY | 1/3 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=True.with_tiered_storage=False.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61370#0194b3d9-1157-45cd-8bee-cd374f2b2f0b | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61370#0194b3d9-1157-45cd-8bee-cd374f2b2f0b | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61462">build#61462</a></summary>


| test_id                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61462#0194be4b-095c-469e-8137-3bb61ddd13ac | FLAKY         | 1/2      |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61462#0194be50-6662-48e6-8797-ba60b9d36fc2 | FLAKY         | 1/4      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61462#0194be50-6662-48e6-8797-ba60b9d36fc2 | FLAKY         | 1/3      |
</details>
<!--/body-->

dotnwat commented on src/v/crash_tracker/report_generator.cc:
```diff
@@ -0,0 +1,92 @@
+
+namespace {}
+
+int main(int ac, char* av[]) {
```
> crash_tracker: Create crash_report_generator
> Created a simple CLI tool that can be used to generate crash reports on
demand.

Is this like a tool for generating test data or something else?

michael-redpanda replied:
yes... i went back and forth on whether or not I wanted to actually add it into the repo and decided to do it to use with tests... i may back it out later if we end up never using it.

dotnwat replied:
you could also consider adding it under `crash_tracker/tests` directory or moving to `tools/` to make its role more obvious

michael-redpanda committed:
dt/crash: Added test to verify offline_log_viewer

Verifies new functionality of parsing crash reports using the
offline_log_viewer tool.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push [dd20b5c](https://github.com/redpanda-data/redpanda/commit/dd20b5c9f0b2fdf23522a8ddbd62062ae5864279):

* Added test for `offline_log_viewer`
* Addressed other PR comments
* Moved report generator to `crash_tracker/tests`

dotnwat commented on src/v/crash_tracker/tests/BUILD:
```diff
@@ -14,3 +15,16 @@ redpanda_cc_gtest(
+
+redpanda_cc_binary(
+    name = "crash_report_generator",
+    srcs = ["report_generator.cc"],
```
@michael-redpanda we should mark this `testonly = true` when you get a chance.

pgellert replied:
Fixed here: https://github.com/redpanda-data/redpanda/pull/25021
